samples/drivers/i2s/output/src/main.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/i2s.h>
+#include <zephyr/sys/iterable_sections.h>
+
+#define SAMPLE_NO 64
+
+/* The data_l represent a sine wave */
+static int16_t data_l[SAMPLE_NO] = {
+	  3211,   6392,   9511,  12539,  15446,  18204,  20787,  23169,
+	 25329,  27244,  28897,  30272,  31356,  32137,  32609,  32767,
+	 32609,  32137,  31356,  30272,  28897,  27244,  25329,  23169,
+	 20787,  18204,  15446,  12539,   9511,   6392,   3211,      0,
+	 -3212,  -6393,  -9512, -12540, -15447, -18205, -20788, -23170,
+	-25330, -27245, -28898, -30273, -31357, -32138, -32610, -32767,
+	-32610, -32138, -31357, -30273, -28898, -27245, -25330, -23170,
+	-20788, -18205, -15447, -12540,  -9512,  -6393,  -3212,     -1,
+};
+
+/* The data_r represent a sine wave shifted by 90 deg to data_l sine wave */
+static int16_t data_r[SAMPLE_NO] = {
+	 32609,  32137,  31356,  30272,  28897,  27244,  25329,  23169,
+	 20787,  18204,  15446,  12539,   9511,   6392,   3211,      0,
+	 -3212,  -6393,  -9512, -12540, -15447, -18205, -20788, -23170,
+	-25330, -27245, -28898, -30273, -31357, -32138, -32610, -32767,
+	-32610, -32138, -31357, -30273, -28898, -27245, -25330, -23170,
+	-20788, -18205, -15447, -12540,  -9512,  -6393,  -3212,     -1,
+	  3211,   6392,   9511,  12539,  15446,  18204,  20787,  23169,
+	 25329,  27244,  28897,  30272,  31356,  32137,  32609,  32767,
+};
+
+
+static void fill_buf(int16_t *tx_block, int att)
+{
+	for (int i = 0; i < SAMPLE_NO; i++) {
+		tx_block[2 * i] = data_l[i] >> att;
+		tx_block[2 * i + 1] = data_r[i] >> att;
```

Comment:
This code assumes that the compiler implementation-defined way of right shifting signed integers is to do sign extension. Zephyr has `arithmetic_shift_right()` for 64-bit integers, not sure if we should have a variant for lower bit widths.

Reply:
AN alternative would be to just replace the right shift with division, i.e. `data_l[i] / (1 << att)`. The rounding direction in division is implementation defined in C89, but Zephyr requires C99 and above that does round towards zero.

Reply:
Thanks for pointing this out- let's use division here instead, that way we avoid possible compiler dependencies.